src/java.base/share/classes/java/text/MessageFormat.java

```diff
@@ -562,60 +654,81 @@ public String toPattern() {
         for (int i = 0; i <= maxOffset; ++i) {
             copyAndFixQuotes(pattern, lastOffset, offsets[i], result);
             lastOffset = offsets[i];
-            result.append('{').append(argumentNumbers[i]);
-            Format fmt = formats[i];
-            if (fmt == null) {
-                // do nothing, string format
-            } else if (fmt instanceof NumberFormat) {
-                if (fmt.equals(NumberFormat.getInstance(locale))) {
-                    result.append(",number");
-                } else if (fmt.equals(NumberFormat.getCurrencyInstance(locale))) {
-                    result.append(",number,currency");
-                } else if (fmt.equals(NumberFormat.getPercentInstance(locale))) {
-                    result.append(",number,percent");
-                } else if (fmt.equals(NumberFormat.getIntegerInstance(locale))) {
-                    result.append(",number,integer");
-                } else {
-                    if (fmt instanceof DecimalFormat) {
-                        result.append(",number,").append(((DecimalFormat)fmt).toPattern());
-                    } else if (fmt instanceof ChoiceFormat) {
-                        result.append(",choice,").append(((ChoiceFormat)fmt).toPattern());
-                    } else {
-                        // UNKNOWN
-                    }
+            result.append('{')
+                    .append(argumentNumbers[i])
+                    .append(patternFromFormat(formats[i]))
+                    .append('}');
+        }
+        copyAndFixQuotes(pattern, lastOffset, pattern.length(), result);
+        return result.toString();
+    }
+
+    /**
+     * This method converts a Format into a {@code FormatType} and {@code
+     * FormatStyle}, if applicable. For each Format, this method will
+     * first check against the pre-defined styles established in the
+     * {@link ##patterns patterns section}. If a Format does not match to a
+     * pre-defined style, it will provide the {@code SubformatPattern}, if the Format
+     * class can provide one. The following subformats do not provide a {@code
+     * SubformatPattern}: {@link CompactNumberFormat}, {@link ListFormat}, and {@link
+     * java.time.format.DateTimeFormatter}. In addition, {@code DateTimeFormatter}
+     * does not implement {@code equals()}, and thus cannot be represented as a
+     * pattern. Any "default"/"medium" styles are omitted according to the specification.
+     */
+    private String patternFromFormat(Format fmt) {
+        if (fmt instanceof NumberFormat) {
+            // Add any instances returned from the NumberFormat factory methods
+            if (fmt.equals(NumberFormat.getInstance(locale))) {
```

Comment:
This looks like wack-a-mole code, no good design; likely to be hard to maintain.
(I don't have a better idea at the moment though).

Reply:
I agree, it was the existing design that caused for example, CompactNumberFormat to not automatically be supported by MessageFormat. A simple alternative would be storing the potential pre-defined NumberFormats in some data structure that we could just iterate through to feel less “whack a mole” like, but that array would still suffer from the same maintenance issues. I’ll try to think of something better.

Reply:
One idea would be to delegate to a (package-private) method in the formatXXX class. 
That would localize to the respective class the details.
(An abstract protected method might be preferred, but its not worth creating extra public API surface area for this).

Reply:
Would it work with custom implementations for say NumberFormat via the SPI?

Reply:
Just moving the if...then... else code to a package-private static method in the respective XXFormat class would work the same.
I had not thought of an addition to the spi to delegate to/through the provider.